public/javascripts/PrairieEditor.js

```diff
@@ -0,0 +1,127 @@
+/**
+ * PrairieEditor is currently a simple jQuery script that sets up ace editors
+ * and configures a save button for each file. It also previews the question by
+ * asynchronously rendering variants and grading submissions.
+ */
+$(function() {
+
+    // Ensure the CSRF token has been set
+    if (! document.csrfToken)
+        console.error('No CSRF token set, asynchronous requests will fail.');
+
+    // Creates an XMLHttpRequest object that posts to the same URL as the editor
+    var post = function(data, onComplete) {
+        var xhr = new XMLHttpRequest();
+        xhr.open("POST", document.URL, true);
+        xhr.setRequestHeader('x-csrf-token', document.csrfToken);
+        xhr.setRequestHeader("Content-Type", "application/json");
+
+        xhr.onreadystatechange = function() {
+            if (xhr.readyState == 4 && xhr.status == 200 && onComplete != null)
+                onComplete(xhr.responseText);
+        };
+
+        if (data)
+            xhr.send(JSON.stringify(data));
+
+        return xhr;
+    };
+
+    // For each editable zone
+    $('[data-edit]').each(function() {
+        // Get file data
+        var fileName = $(this).data('edit');
+        var fileMode = $(this).data('mode');
+
+        // Get IDs of editor and save button
+        var editorId = $(this).attr('id');
+        var saveId = $(this).data('save');
+        var saveButton = $('#' + saveId);
+
+        // Create an Ace editor instance
+        var editor = ace.edit(editorId);
+        editor.setTheme('ace/theme/monokai');
+        editor.getSession().setMode('ace/mode/' + fileMode);
+
+        // Enable the save button if the content is different from the current value
+        var currentContent = editor.getValue();
+        saveButton.attr('disabled', 'true');
+        editor.getSession().on('change', function() {
+            saveButton.attr('disabled', editor.getValue() == currentContent);
+        });
+
+        // When the save button is clicked
+        saveButton.click(function() {
+
+            // asynchronously save the file content
+            post({
+                action: 'save',
+                file: fileName,
+                content: editor.getValue()
+            },
+            function(response) {
+                // Reset the save button to be disabled (until the next edit)
+                saveButton.attr('disabled', true);
+                if (response != 'error') {
+                    currentContent = editor.getValue();
+
+                    // Refresh the question preview
+                    generatePreview();
```

Comment:
I wonder whether we should preserve the current preview question variant when saving. For example, if someone is editing the HTML to affect the correct answer panel, it will be annoying if it keeps disappearing every time they save. On the other hand, if we make a large change to a question then it will be annoying to have the preview break because the code is now totally different. Maybe it's best to leave things as they are.